try-to-fix-two-linux-kernel-bugs-while-testing-tidb-operator-in-k8s.md

```diff
@@ -0,0 +1,188 @@
+
+If this operation failed, the kernel fell back to allocate the `bio_vec` object from the mempool. However, inside the mempool, the system first tried to execute the `pool->alloc` callback for allocation. If this allocation failed, the kernel set the process to an uninterruptible state and put the process in a wait queue. After other processes returned memory resources to the mempool or the timer timed out (after 5 seconds), the process scheduler invoked the process for retry. The wait time of the wait queue was consistent with the jitter delay of our application monitoring.
+
+When we created the Docker container, we didn't set `memory.kmem.limit_in_bytes`. But why was kmem insufficient? To determine whether `memory.kmem.limit_in_bytes` was set, we went to the cgroup memory controller to check the container's kmem information. The kmem statistics were enabled, and the kemem limit was set to a very large value.   
+
+Because we knew that the kmem accounting was unstable in the RHEL 3.10 kernel, we suspected that a kernel bug caused the SLUB allocation failure. We searched for kernel patch information, and found that it was a kernel bug, and that it had been fixed in Linux kernel version 4.x: [slub: make dead caches discard free slabs immediately](https://github.com/torvalds/linux/commit/d6e0b7fa11862433773d986b5f995ffdf47ce672). There was also a namespace leak issue associated with kmem accounting: [mm: memcontrol: fix cgroup creation failure after many small jobs](https://github.com/torvalds/linux/commit/73f576c04b9410ed19660f74f97521bee6e1c546).
+
+So how was kmem accounting enabled? We used the [opensnoop](https://github.com/iovisor/bcc/blob/master/tools/opensnoop.py) tool in [bcc](http://github.com/iovisor/bcc) to monitor the kmem configuration file and captured `[runc](https://github.com/opencontainers/runc)` as the file modifier. From the K8s code, we found that kmem accounting was enabled by default in the K8s-dependent `runc` project. 
+
+### Solution
+
+Based on our issue analysis, we can fix the bug in either of the following ways:
+
+- Upgrade the existing kernel to a later version.
+- Disable the kmem accounting feature when starting the container.
+
+Now `runc` includes the conditional compilation option, and you can disable kmem accounting via [Build Tags](https://github.com/opencontainers/runc#build-tags). After we disabled this feature, our test result showed that the jitter disappeared, and so did the namespace leak and SLUB allocation failure problems.
+
+#### Operation steps
+
+We need to disable the kmem accounting feature on both kubelet and Docker.
+
+1. Recompile kubelet according to the corresponding kubelet version:
+
+    - For kubelet v1.14 or later, add [Build Tags](https://github.com/kubernetes/kubernetes/blob/release-1.14/vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs/kmem_disabled.go#L1) when you compile kubelet to disable kmem accounting:
+
+    ```
+    $ make BUILDTAGS=”nokmem”
```

Comment:
🤔 should use straight quotes here

```suggestion
    $ make BUILDTAGS="nokmem"
```

Reply:
@kennytm Thanks. 
I use straight quotes here but you see curly quotes. So strange. But we can see straight quotes here on the PingCAP website :)